### PR TITLE
[DynamoDB] StateStore - Add TTL from metadata to Set and BulkSet

### PR DESCRIPTION
**Pull request by wimvdc**
# Description

Adding ttlInSeconds support to DynamoDB as state store (Set and SetBulk)
This is my first ever experience with Go, so please let me know how to improve the new code.

## Issue reference

#1489

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: state/aws/dynamodb/dynamodb.go
Comment (berndverst) [inline state/aws/dynamodb/dynamodb.go:309]:
```suggestion
		//DynamoDB expects an epoch timestamp in seconds
```

File: state/aws/dynamodb/dynamodb.go
Comment (berndverst) [inline state/aws/dynamodb/dynamodb.go:310]:
```suggestion
		expirationTime := time.Now().Unix() + parsedInt
```

File: state/aws/dynamodb/dynamodb.go
Comment (berndverst) [inline state/aws/dynamodb/dynamodb.go:312]:
```suggestion
		return &expirationTime, nil
```

File: state/aws/dynamodb/dynamodb_test.go
Comment (berndverst) [inline state/aws/dynamodb/dynamodb_test.go:316]: Can you please update every occurrence of `un-successfully` to be `unsuccessfully`?

**Review by berndverst — CHANGES_REQUESTED** [review]

**Comment by berndverst** [comment]
Also please fix linter errors!

File: state/aws/dynamodb/dynamodb.go
Comment (berndverst) [inline state/aws/dynamodb/dynamodb.go:306]: `-1` was intended to be used to explicitly persist a record, for example in the case where the state store supports a default TTL that is applied to every new entity.

It does not seem that DynamoDB supports such a concept. Perhaps you should return for every negative value?

**Comment by berndverst** [comment]
From my understanding of the Dynamo DB documentation the attribute used for TTL is configurable when TTL is enabled on a table. https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-before-you-start.html

However this PR here uses a hardcoded `expireAt`.

I think this should be configurable via the component Metadata, similar to how the table name is being specified today:
https://github.com/dapr/components-contrib/blob/master/state/aws/dynamodb/dynamodb.go#L42

Can you please update the PR accordingly? Also make sure to create a follow up PR against the Dapr Docs to document the TTL feature, and document this component metadata field to specify the TTL attribute (and that someone must manually enable the TTL feature and for the Dynamo DB table and specify the attribute name)

Thanks!

**Comment by berndverst** [comment]
Another thought:
The Dynamo DB documentation suggests that entities that have expired are still returned for 48 hours or so.

Should the Get & Bulk Get operations examine the current expiryTime and filter out those entities if that expiryTime is in the past?

**Comment by wimvdc** [comment]
> From my understanding of the Dynamo DB documentation the attribute used for TTL is configurable when TTL is enabled on a table. https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-before-you-start.html or https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-how-to.html
> 
> However this PR here uses a hardcoded `expireAt`. I think this should be configurable via the component Metadata, similar to how the table name is being specified today: https://github.com/dapr/components-contrib/blob/master/state/aws/dynamodb/dynamodb.go#L42
> 

I have changed it so that the component metadata needs to have `ttlAttributeName` specified in order to make use of TTL.

> And of course this component metadata field (as opposed to the state store request metadata) should be optional. If the field is not provided, then the TTL feature cannot function, so any state store request TTL values should be ignored.
> 
I added an extra if in the `parseTTL` function

> Can you please update the PR accordingly? Also make sure to create a follow up PR against the Dapr Docs to document the TTL feature, and document this component metadata field to specify the TTL attribute (and that someone must manually enable the TTL feature and for the Dynamo DB table and specify the attribute name)
> 
> This is the page that needs to be updated with the optional metadata field and more details: https://docs.dapr.io/reference/components-reference/supported-state-stores/setup-dynamodb/ And you'll want to update the following table to check the TTL column: https://docs.dapr.io/reference/components-reference/supported-state-stores/
> 
> Thanks!

I'm not sure what the order should be, wait until this PR has been merged? Or in parallel? 

> Another thought: The Dynamo DB documentation suggests that entities that have expired are still returned for 48 hours or so.
> 
> Should the Get & Bulk Get operations examine the current expiryTime and filter out those entities if that expiryTime is in the past?
That's a hard one. 

On the one hand you have chosen to use DynamoDB in favor of another technology/service as state store so I might be strange to temper with the functionalities that DynamoDB offers. 

On the other hand developers might not be aware of this behaviour, especially if they try to develop cloud agnostic. 

Personally I would not implement the extra checks in (bulk) Get. But it might be useful to return the TTL field in (bulk) Get so that a developer can choose to do something with it ?

File: state/aws/dynamodb/dynamodb.go
Comment (wimvdc) [reply state/aws/dynamodb/dynamodb.go]: What I tried to archive;
Imagine you create the following item:
```
key: dummyKey
value: someValue
ttl: 90
```

And 20 seconds later you do:
```
key: dummyKey
value: someValue
```

Will this override the existing TTL field so that DynamoDB won't delete it after 70 seconds?

Comment (berndverst) [reply state/aws/dynamodb/dynamodb.go]: That depends on how Dynamo DB handles this (which I personally don't know as I am not a user of DynamoDB). In general I think that should indeed reset the TTL to whatever the default is (no TTL?).

If there is a default TTL (which I think Dynamo DB doesn't have) you might need to specify explicitly that you want to ignore the default TTL and persist the item. That's what `-1` was intended to be for.

I think you don't want have to worry about that -- I can't seem to find anything about default TTLs / globally defined TTLs in Dynamo DB, so it probably doesn't apply here.

**Comment by berndverst** [comment]
image

**Comment by berndverst** [comment]
@wimvdc discussion here:
https://discord.com/channels/778680217417809931/781589820128493598/942871929306636428

@yaron2 and I agree that BulkGet and Get should filter out all records with expired TTL in the DynamoDB state store component.

Thanks!

**Comment by wimvdc** [comment]
> > > Also please fix linter errors!
> > 
> > 
> > Should be fixed now
> > > From my understanding of the Dynamo DB documentation the attribute used for TTL is configurable when TTL is enabled on a table. https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-before-you-start.html or https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-how-to.html
> > > However this PR here uses a hardcoded `expireAt`. I think this should be configurable via the component Metadata, similar to how the table name is being specified today: https://github.com/dapr/components-contrib/blob/master/state/aws/dynamodb/dynamodb.go#L42
> > 
> > 
> > I have changed it so that the component metadata needs to have `ttlAttributeName` specified in order to make use of TTL.
> > > And of course this component metadata field (as opposed to the state store request metadata) should be optional. If the field is not provided, then the TTL feature cannot function, so any state store request TTL values should be ignored.
> > 
> > 
> > I added an extra if in the `parseTTL` function
> > > Can you please update the PR accordingly? Also make sure to create a follow up PR against the Dapr Docs to document the TTL feature, and document this component metadata field to specify the TTL attribute (and that someone must manually enable the TTL feature and for the Dynamo DB table and specify the attribute name)
> > > This is the page that needs to be updated with the optional metadata field and more details: https://docs.dapr.io/reference/components-reference/supported-state-stores/setup-dynamodb/ And you'll want to update the following table to check the TTL column: https://docs.dapr.io/reference/components-reference/supported-state-stores/
> > > Thanks!
> > 
> > 
> > I'm not sure what the order should be, wait until this PR has been merged? Or in parallel?
> > > Another thought: The Dynamo DB documentation suggests that entities that have expired are still returned for 48 hours or so.
> > > Should the Get & Bulk Get operations examine the current expiryTime and filter out those entities if that expiryTime is in the past?
> > > That's a hard one.
> > 
> > 
> > On the one hand you have chosen to use DynamoDB in favor of another technology/service as state store so I might be strange to temper with the functionalities that DynamoDB offers.
> > On the other hand developers might not be aware of this behaviour, especially if they try to develop cloud agnostic.
> > Personally I would not implement the extra checks in (bulk) Get. But it might be useful to return the TTL field in (bulk) Get so that a developer can choose to do something with it ?
> 
> @wimvdc in general when you look at Dapr components you'll find that the state store components are the fully managed components that hide a lot of details from you. Whenever you want more low level control you use the equivalent bindings.
> 
> My vote would be to filter out the expired keys and to perform these extra checks.
> 
> If you want this complete DynamoDB control I would ensure (in a separate PR and it doesn't have to be done by you) that the Dynamo DB binding instead has a way to return all entities including the TTL state. https://docs.dapr.io/reference/components-reference/supported-bindings/dynamodb/
> 
> What do you think?



> @wimvdc discussion here: https://discord.com/channels/778680217417809931/781589820128493598/942871929306636428
> 
> @yaron2 and I agree that BulkGet and Get should filter out all records with expired TTL in the DynamoDB state store component.
> 
> Thanks!

Sounds reasonable, will work on the filter

File: state/aws/dynamodb/dynamodb.go
Comment (wimvdc) [reply state/aws/dynamodb/dynamodb.go]: Ok, I removed the IF statement and modified the tests a bit.

**Comment by wimvdc** [comment]
I'm currently stuck with DCO-stuff.
I need to signoff some old commits but I can't use rebase because that would override merged commits.
Any help?

**Comment by daixiang0** [comment]
I think you can rebase main branch and force push to fix DCO.

**Comment by wimvdc** [comment]
I really messed up the git history.
Closing this in favor of a new PR with correct commits, sorry for my unorthodox way of working.
New PR #1512

**Pull request closed without merging**
